lib/srv/reexec.go

```diff
@@ -612,9 +612,77 @@ func (o *osWrapper) startNewParker(ctx context.Context, credential *syscall.Cred
 	return nil
 }
 
-// RunForward reads in the command to run from the parent process (over a
+type forwardHandler func(ctx context.Context, addr string, file *os.File) error
+
+func handleLocalPortForward(ctx context.Context, addr string, file *os.File) error {
+	conn, err := uds.FromFile(file)
+	_ = file.Close()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	defer conn.Close()
+	var d net.Dialer
+	remote, err := d.DialContext(ctx, "tcp", addr)
+	if err != nil {
+		log.WithError(err).Errorf("error dialing %v", addr)
+	}
+	defer remote.Close()
+	if err := utils.ProxyConn(ctx, conn, remote); err != nil {
+		log.WithError(err).Errorf("error proxying connection to %v", addr)
+	}
+
+	return nil
+}
+
+func createRemotePortForwardingListener(ctx context.Context, addr string) (*os.File, error) {
+	var lc net.ListenConfig
+	listener, err := lc.Listen(ctx, "tcp", addr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer listener.Close()
+
+	tcpListener, _ := listener.(*net.TCPListener)
+	if tcpListener == nil {
+		return nil, trace.Errorf("expected listener to be of type *net.TCPListener, but was %T", listener)
+	}
+
+	listenerFD, err := tcpListener.File()
+	return listenerFD, trace.Wrap(err)
+}
+
+func handleRemotePortForward(ctx context.Context, addr string, file *os.File) error {
+	controlConn, err := uds.FromFile(file)
+	_ = file.Close()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	// unblock the final write
+	context.AfterFunc(ctx, func() { _ = controlConn.Close() })
+	go func() {
+		defer cancel()
+		_, _ = controlConn.Read(make([]byte, 1))
+	}()
+
+	listenerFD, err := createRemotePortForwardingListener(ctx, addr)
+	var payload []byte
+	var files []*os.File
```

Comment:
```suggestion
	var payload []byte
	var files []*os.File
	listenerFD, err := createRemotePortForwardingListener(ctx, addr)
```

Just to keep the `err` assignment and the error check as close together as possible.